react-ui/src/pages/contest/ContestAppBar.jsx

```diff
@@ -0,0 +1,112 @@
+/**
+ * Contest-specific content for top app bar
+ */
+
+import Box from '@material-ui/core/Box';
+import EmojiEventsOutlinedIcon from '@material-ui/icons/EmojiEventsOutlined';
+import SettingsOutlinedIcon from '@material-ui/icons/SettingsOutlined';
+import ThumbsUpDownOutlinedIcon from '@material-ui/icons/ThumbsUpDownOutlined';
+import PropTypes from 'prop-types';
+import { useLocation, useParams } from 'react-router-dom';
+
+import {
+  useSwrData,
+} from '../../common';
+import {
+  ArrowBackButton,
+  CustomIconButton,
+  Countdown,
+} from '../../components';
+
+function ContestAppBarMain({ handleVotingExpired, handleReload }) {
+  const { contestId } = useParams();
+  const { state = {} } = useLocation();
+
+  const apiPath = `/contests/${contestId}`;
+  const { data: contest } = useSwrData(apiPath, false);
+
+  const backLink = (state || {}).back || '/contests';
+
+  const {
+    date,
+    isContestMode,
+    voteEnd,
+  } = contest;
+  const voteEndDate = new Date(voteEnd);
+
+  return (
+    <>
+      <ArrowBackButton state={{ date }} to={backLink} />
+      {isContestMode && (
+        <Box display="inline-flex" paddingLeft={1.5}>
+          <Countdown
+            endDate={voteEndDate}
+            handleExpiry={handleVotingExpired}
+            handleReload={handleReload}
+          />
+        </Box>
+      )}
+    </>
+  );
+}
+
+function ContestAppBarRight({ setInfoOpen }) {
+  const { contestId } = useParams();
+
+  const apiPath = `/contests/${contestId}`;
+  const { data: contest } = useSwrData(apiPath, false);
+
+  const toggleSettingsOpen = () => {
+    setInfoOpen((prev) => !prev);
+  };
+
+  const {
+    localVoting,
+    validRedditId,
+    winnersThreadId,
+  } = contest;
+
+  return (
+    <>
+      {!localVoting && validRedditId && (
+        <CustomIconButton
+          href={`https://redd.it/${contestId}`}
+          ariaLabel="Open voting thread"
+          Icon={ThumbsUpDownOutlinedIcon}
+        />
+      )}
+      {winnersThreadId && (
+        <CustomIconButton
+          href={`https://redd.it/${winnersThreadId}`}
+          ariaLabel="Open winners thread"
+          Icon={EmojiEventsOutlinedIcon}
+        />
+      )}
+      <CustomIconButton
+        ariaLabel="View settings"
+        onClick={toggleSettingsOpen}
+        Icon={SettingsOutlinedIcon}
+      />
+    </>
+  );
+}
+
+ContestAppBarRight.propTypes = {
+  setInfoOpen: PropTypes.func,
+};
+
+ContestAppBarRight.defaultProps = {
+  setInfoOpen: () => { },
+};
+
+ContestAppBarMain.propTypes = {
+  handleVotingExpired: PropTypes.func,
+  handleReload: PropTypes.func,
```

Comment:
Is there a reason these `func`s aren't all defined as `isRequired`? These functions are always going to passed in and they're expected to be there.

Reply:
This comment extends to several of the other files as well. Setting props, especially functions, to default values may make errors a bit more difficult to find. E.g. the function being passed in is accidentally deleted and instead of seeing a warning in the console during development mode, you don't notice but the action never happens because there's a no-op function being defaulted. I'd prefer a page possibly breaking rather than silently failing and it's not clear that something is wrong (I'd actually prefer some try-catch or a check before a function call, but that's also tedious).

Reply:
Got it, I'll hunt and correct.

Reply:
Got it, see comment below for appBar